lib/ykpiv.c

```diff
@@ -246,95 +274,119 @@ ykpiv_rc ykpiv_list_readers(ykpiv_state *state, char *readers, size_t *len) {
+  if((rc & 0xFFFFFFFF) == SCARD_W_RESET_CARD) {
+    if((res = _reconnect(state)) != YKPIV_OK) {
+      return res;
+    }
+    return _begin_transaction(state);
+  }
   if(rc != SCARD_S_SUCCESS) {
     if(state->verbose) {
       fprintf(stderr, "error: Failed to begin pcsc transaction, rc=%08lx\n", rc);
     }
     return YKPIV_PCSC_ERROR;
   }
-  do {
-    size_t this_size = 0xff;
-    unsigned char data[261];
-    unsigned long recv_len = sizeof(data);
-    APDU apdu;
-
-    memset(apdu.raw, 0, sizeof(apdu.raw));
-    memcpy(apdu.raw, templ, 4);
-    if(in_ptr + 0xff < in_data + in_len) {
-      apdu.st.cla = 0x10;
-    } else {
-      this_size = (size_t)((in_data + in_len) - in_ptr);
-    }
-    if(state->verbose > 2) {
-      fprintf(stderr, "Going to send %lu bytes in this go.\n", (unsigned long)this_size);
-    }
-    apdu.st.lc = this_size;
-    memcpy(apdu.st.data, in_ptr, this_size);
-    res = send_data(state, &apdu, data, &recv_len, sw);
-    if(res != YKPIV_OK) {
-      return res;
-    } else if(*sw != SW_SUCCESS && *sw >> 8 != 0x61) {
-      return YKPIV_OK;
-    }
-    if(*out_len + recv_len - 2 > max_out) {
-      if(state->verbose) {
-  fprintf(stderr, "Output buffer to small, wanted to write %lu, max was %lu.\n", *out_len + recv_len - 2, max_out);
-      }
-      return YKPIV_SIZE_ERROR;
-    }
-    if(out_data) {
-      memcpy(out_data, data, recv_len - 2);
-      out_data += recv_len - 2;
-      *out_len += recv_len - 2;
-    }
-    in_ptr += this_size;
-  } while(in_ptr < in_data + in_len);
-  while(*sw >> 8 == 0x61) {
-    APDU apdu;
-    unsigned char data[261];
-    unsigned long recv_len = sizeof(data);
-
-    if(state->verbose > 2) {
-      fprintf(stderr, "The card indicates there is %d bytes more data for us.\n", *sw & 0xff);
-    }
+  return YKPIV_OK;
+}
 
-    memset(apdu.raw, 0, sizeof(apdu.raw));
-    apdu.st.ins = 0xc0;
-    res = send_data(state, &apdu, data, &recv_len, sw);
-    if(res != YKPIV_OK) {
-      return res;
-    } else if(*sw != SW_SUCCESS && *sw >> 8 != 0x61) {
-      return YKPIV_OK;
-    }
-    if(*out_len + recv_len - 2 > max_out) {
-      fprintf(stderr, "Output buffer to small, wanted to write %lu, max was %lu.", *out_len + recv_len - 2, max_out);
-    }
-    if(out_data) {
-      memcpy(out_data, data, recv_len - 2);
-      out_data += recv_len - 2;
-      *out_len += recv_len - 2;
-    }
-  }
-  rc = SCardEndTransaction(state->card, SCARD_LEAVE_CARD);
-  if(rc != SCARD_S_SUCCESS) {
-    if(state->verbose) {
-      fprintf(stderr, "error: Failed to end pcsc transaction, rc=%08lx\n", rc);
-    }
+static ykpiv_rc _end_transaction(ykpiv_state *state) {
+  long rc = SCardEndTransaction(state->card, SCARD_LEAVE_CARD);
+  if(rc != SCARD_S_SUCCESS && state->verbose) {
+    fprintf(stderr, "error: Failed to end pcsc transaction, rc=%08lx\n", rc);
     return YKPIV_PCSC_ERROR;
   }
   return YKPIV_OK;
 }
 
+ykpiv_rc ykpiv_transfer_data(ykpiv_state *state, const unsigned char *templ,
```

Comment:
Please fix the indentation of this function to match the rest of the file, essentially adding two spaces to the beginning of every line.